src/store/actions/auth.js

```diff
@@ -130,13 +149,94 @@ export const authGo = (email, password, isRegister) => {
   }
 };
 
+export const authSetProfileGo = (token, displayName, photoUrl, isEdit) => {
+  return (dispatch) => {
+    const apiKey = process.env.REACT_APP_FIREBASE_API_KEY;
+
+    const authData = {
+      idToken: token,
+      displayName: displayName,
+      photoUrl: photoUrl
+    }
+
+    let url = `https://identitytoolkit.googleapis.com/v1/accounts:update?key=${apiKey}`;
+
+    axios.post(url, authData)
```

Comment:
Personally, I prefer to name my callbacks, and define them outside of the context where they are used, whenever possible:

```
const handleAuthSetProfileGoPostSuccess = ({ dispatch, isEdit, token }) => (response) => {
        dispatch(authDoneLoading());
        dispatch(closeModal());
        setTimeout(() => {
          dispatch(openAndSetActiveModalAndMessage('success', `Your profile has been ${isEdit ? 'updated' : 'set'}!`));
        }, 250);
        dispatch(authGetProfile(token));
      }

const handleAuthSetProfileGoPostFailure = (dispatch) => (error) => {
        const errorMessage = setErrorMessage(error);

         dispatch(authFail(errorMessage));
        dispatch(closeModal());
        setTimeout(() => {
          dispatch(openAndSetActiveModalAndMessage('error', errorMessage));
        }, 250);
      }

export const authSetProfileGo = ... {
...
axios.post(url, authData)
  .then(handleAuthSetProfileGoPostSuccess({ dispatch, isEdit, token }))
  .catch(handleAuthSetProfileGoPostFailure(dispatch))
...
```

Please do note here that we are utilizing the concept of a higher order function in order to get the callbacks what they require in their scope to operate. The success callback needs `dispatch`, `isEdit`, and `token`, so first pass those variables in through a single invocation, then when the `.then` runs, it calls the second function. It could be more clearly written as follows:

```
export const authSetProfileGo = ... {
...
const handleAuthSetProfileGoPostSuccessWithContext = handleAuthSetProfileGoPostSuccess({ dispatch, isEdit, token })

axios.post(url, authData)
  .then(handleAuthSetProfileGoPostSuccessWithContext)
  .catch(handleAuthSetProfileGoPostFailure(dispatch))
...
```

At the definition

```const handleAuthSetProfileGoPostSuccess = ({ dispatch, isEdit, token }) => (response) =>...```

 we are returning a function within the first function, which is what makes the line 

```const handleAuthSetProfileGoPostSuccessWithContext = handleAuthSetProfileGoPostSuccess({ dispatch, isEdit, token })```

make sense. 

By following this sort of pattern, when you're inside `authProfileSetGo` you can clearly understand what's happening in the promise, and what each callback actually needs to operate. It becomes clear that the `.catch` is only `dispatch`ing something from the error it will receive via Axios, and that the `.then` needs some other contexts in order to operate. All of a sudden, your code is more understandable, and if you need to debug something, you don't need to let your eyes and brain be concerned with multiple contexts, when you only need one at a time.

Say something's wrong with the success handler. You then read your `axios` invocation, and see the aptly named `handleAuthSetProfileGoPostSuccess`. You read the arguments which it receives and subsequently go to the definition. You check the arguments list to see if it matches, and then you can actually debug the implementation _away from where you call it_. Writing and tracing code in this manner is much easier from a mental model standpoint and actually leads to much fewer bugs better reasoning in general.

Reply:
@shoelessness I've read this over a few times and it's making sense. The only thing I'm a little iffy about is that I'm doing this same sort of thing all of the time and I'm going to have so many new functions because of it. I guess they'll all be labelled which is good, but isn't the catch always for catching errors. I don't know, I'm just thinking this is more confusing but maybe I'm missing something haha.

Reply:
You will have many new functions. The glorious part is that they don't have to be defined in the same file. Having a `something.helpers.js` or `something.utils.js` means you have a separate place to read the implementation and work on it. Debugging this actually becomes wonderfully straightforward, as the stack trace for an error or log will show you the name of each function in the call stack, which means you don't have to do anything other than look for the name of the thing which is breaking. Making one function do one thing at this point is the key to wonderful code, because then you only need to fix one thing at a time, which is usually where bugs occur in the first place.

The other thing you get is repeatable use cases. Say you thought that all your axios catches were different in significant ways, but then you start to abstract all of them into named functions. As you read through those named functions, you might see commonalities emerge, and a way to DRY the code up. DRY doesn't always work, but sometimes you're doing the same thing over and over and not realizing that you could refactor that block to be more elegant or efficient or whatever, simply because you aren't reading it in one common place.

That said, not everyone enjoys a functional programming approach. As for me, I have found it to be a key to success.

Reply:
@shoelessness Ok that makes sense. I'm just more comfortable right now leaving it as one so I can see what's going on without having to search around for the right function. For now I might just leave it for ease of reference, but I'll definitely try your way the next time. I like the idea of keeping it DRY and reusing so it'll be interesting to see how that goes.